e2e/kind.sh

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env bash
+KUBECONFIG="kind.yaml"
+KIND_CLUSTER="kind-cluster-kilo"
+KIND_BINARY="${KIND_BINARY:-kind}"
+KUBECTL_BINARY="${KUBECTL_BINARY:-kubectl}"
+KILO_IMAGE="${KILO_IMAGE:-squat/kilo}"
+
+is_ready() {
+	for pod in $(${KUBECTL_BINARY} -n ${1} get pods -o name -l ${2}); do
+		${KUBECTL_BINARY} -n ${1} get $pod | tail -n 1 | grep -q  Running;
+		if [ $? -ne 0 ]; then
+			return 1;
+		fi
+	done
+	return 0
+}
+
+# Returns non zero if one pod of the given name in the given namespace is not ready.
+block_until_ready_by_name() {
+	block_until_ready ${1} app.kubernetes.io/name=${2}
+}
+
+# Blocks until all pods of a deployment are ready.
+block_until_ready() {
+	# Just abort after 150s
+	for c in {1..30}; do
+		ready=$(is_ready ${1} ${2})
+		if [ $? -ne 0 ]; then
+			echo "some ${2} pods are not ready, yet. Retries=$c/30"
+			sleep 5
+		else
+			break
+		fi
+	done
+	return 0
+}
+
+# Block waits until pods are ready. When patching pods, it is not very reliable because sometimes it checkts the state of old pods.
+block() {
+	$KUBECTL_BINARY -n ${1} wait -l "app.kubernetes.io/name=${2}" pod  --for=condition=Ready
+}
+
+# Set up the kind cluster and deploy Kilo, Adjacency and a helper with curl.
+setup_suite() {
+	$KIND_BINARY delete clusters $KIND_CLUSTER > /dev/null
+	# Create the kind cluster.
+	$KIND_BINARY create cluster --name $KIND_CLUSTER --config ./kind-config.yaml
+	# Load the Kilo image into kind.
+	docker tag $KILO_IMAGE squat/kilo:test
+	$KIND_BINARY load docker-image squat/kilo:test --name $KIND_CLUSTER
+	# Apply Kilo the the cluster.
+	$KUBECTL_BINARY apply -f kilo-kind-userspace.yaml
+	block_until_ready_by_name kube-system kilo-userspace 
+	$KUBECTL_BINARY wait nodes --all --for=condition=Ready
+	# wait for coredns
+	block_until_ready kube_system k8s-app=kube-dns
+	$KUBECTL_BINARY taint node $KIND_CLUSTER-control-plane node-role.kubernetes.io/master:NoSchedule-
+	$KUBECTL_BINARY apply -f https://raw.githubusercontent.com/heptoprint/adjacency/master/example.yaml
+	$KUBECTL_BINARY apply -f helper-curl.yaml
+	block_until_ready_by_name adjacency adjacency
+	block_until_ready_by_name default curl
+}
+
+block_until_ping() {
+	for c in {1..30}; do
+		keepgoing=1
+		# Block until all IP addresses of the adjacency pods were reached.
+		for ip in $($KUBECTL_BINARY get pods -l app.kubernetes.io/name=adjacency -o jsonpath='{.items[*].status.podIP}'); do
+			ping=$($KUBECTL_BINARY get pods -l app.kubernetes.io/name=curl -o name | xargs -i $KUBECTL_BINARY exec {} -- /bin/sh -c 'curl -s http://'$ip':8080/ping')
+			if [[ $ping == "pong" ]]; then
+				keepgoing=0
+			else
+				keepgoing=1
+				echo "expected \"pong\" got \"$ping\""
+				break
+			fi
+		done
```

Comment:
I added this. I think sometimes only one pod was ready. So now all IPs of the service must be up.